diff --git src/ngx_http_lua_util.c src/ngx_http_lua_util.c
index 8fd26561..727ca3da 100644
--- src/ngx_http_lua_util.c
+++ src/ngx_http_lua_util.c
@@ -599,6 +599,12 @@ ngx_http_lua_send_chain_link(ngx_http_request_t *r, ngx_http_lua_ctx_t *ctx,
     if (r->header_only) {
         ctx->eof = 1;
 
+        if (!r->request_body && r == r->main) {
+            if (ngx_http_discard_request_body(r) != NGX_OK) {
+                return NGX_ERROR;
+            }
+        }
+
         if (ctx->buffering) {
             return ngx_http_lua_send_http10_headers(r, ctx);
         }
